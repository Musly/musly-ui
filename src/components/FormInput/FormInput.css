.input {
  display: flex;

  &:hover,
  &:focus-within {
    .input__label {
      @apply text-gray-600;
    }

    .input__label--secondary {
      @apply text-gray-400;
    }

    .input__label--error {
      @apply text-red-500;
    }

    .input__input {
      @apply border-blue-300;
    }

    .input__input--secondary {
      @apply border-yellow-300;
    }

    .input__input--error {
      @apply border-red-500;
    }
  }

  &--horizontal {
    flex-direction: row;
    align-items: center;
  }

  &--vertical {
    flex-direction: column;
  }
}

.input__label {
  font-weight: 600;
  cursor: pointer;
  transition-property: color;
  transition-duration: 300ms;

  @apply mb-1 ease-sharp text-gray-500 text-sm uppercase;
}

.input__label--horizontal {
  width: 20rem;
}

.input__label--secondary {
  @apply text-gray-300;
}

.input__label--error {
  @apply text-red-500;
}

.input__input {
  width: 100%;

  @apply text-black form-input outline-none focus:outline-none focus:ring-2 border
    ring-opacity-20 bg-white py-3 px-4 rounded-md transition duration-200 ease-sharp
    ring-blue-500 border-gray-300 focus:border-blue-500;

  &::placeholder {
    @apply text-gray-400 ease-sharp;
  }
}

.input__input--secondary {
  @apply ring-opacity-40 ring-yellow-500 border-gray-100 focus:border-yellow-500;
}

.input__input--error {
  @apply ring-opacity-20 ring-red-500 border-red-500 focus:border-red-500;
}

.input__error {
  @apply text-red-500 mt-1 text-sm;
}
