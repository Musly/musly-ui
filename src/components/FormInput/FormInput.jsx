/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import './FormInput.css';

export function FormInput (props) {
  const {
    horizontal, error, secondary, label, value, onChange, placeholder, ...rest
  } = props;

  const wrapperClasses = classNames({
    input: true,
    'input--horizontal': horizontal,
    'input--vertical': !horizontal,
  });

  const labelClasses = classNames({
    input__label: true,
    'input__label--error': !!error,
    'input__label--secondary': !error && secondary,
    'input__label--horizontal': horizontal,
  });

  const inputClasses = classNames({
    input__input: true,
    'input__input--error': !!error,
    'input__input--secondary': !error && secondary,
  });

  return (
    <label className={wrapperClasses}>
      {label && (
        <span className={labelClasses}>
          {label}
        </span>
      )}

      <input
        type="text"
        {...rest}
        value={value}
        onChange={onChange}
        placeholder={placeholder}
        className={inputClasses}
      />

      {error && (
        <div className="input__error">
          {error}
        </div>
      )}
    </label>
  );
}

FormInput.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  label: PropTypes.string,
  secondary: PropTypes.bool,
  horizontal: PropTypes.bool,
  error: PropTypes.string,
  placeholder: PropTypes.string,
};

FormInput.defaultProps = {
  onChange: null,
  label: null,
  secondary: false,
  horizontal: false,
  error: null,
  placeholder: null,
};
