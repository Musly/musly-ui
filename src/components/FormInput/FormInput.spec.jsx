import React from 'react';
import { shallow } from 'enzyme';
import { FormInput } from './FormInput';

describe('<FormInput />', () => {
  it('should render without label and error', () => {
    const onChange = jest.fn();
    const result = shallow(<FormInput value="" onChange={onChange} />);
    expect(result.find('.input__input')).toExist();
    expect(result.find('.input__label')).not.toExist();
    expect(result.find('.input__error')).not.toExist();
    result.find('.input__input').simulate('change');
    expect(onChange).toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });

  it('should render with label', () => {
    const result = shallow(<FormInput value="" label="Test label" />);
    expect(result.find('.input__input')).toExist();
    expect(result.find('.input__label')).toExist();
    expect(result.find('.input__error')).not.toExist();
    expect(result).toMatchSnapshot();
  });

  it('should render with error', () => {
    const result = shallow(<FormInput value="" label="Test label" error="Test error" />);
    expect(result.find('.input__input')).toExist();
    expect(result.find('.input__label')).toExist();
    expect(result.find('.input__error')).toExist();
    expect(result).toMatchSnapshot();
  });
});
