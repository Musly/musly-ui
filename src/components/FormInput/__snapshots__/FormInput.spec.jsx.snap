// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormInput /> should render with error 1`] = `
<label
  className="input input--vertical"
>
  <span
    className="input__label input__label--error"
  >
    Test label
  </span>
  <input
    className="input__input input__input--error"
    onChange={null}
    placeholder={null}
    type="text"
    value=""
  />
  <div
    className="input__error"
  >
    Test error
  </div>
</label>
`;

exports[`<FormInput /> should render with label 1`] = `
<label
  className="input input--vertical"
>
  <span
    className="input__label"
  >
    Test label
  </span>
  <input
    className="input__input"
    onChange={null}
    placeholder={null}
    type="text"
    value=""
  />
</label>
`;

exports[`<FormInput /> should render without label and error 1`] = `
<label
  className="input input--vertical"
>
  <input
    className="input__input"
    onChange={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    placeholder={null}
    type="text"
    value=""
  />
</label>
`;
